lib/scope.js

```diff
@@ -633,167 +623,213 @@ module.exports = {
-      return {
-        disabled: elem.disabled,
-        tag: elem.tagName,
-        type: elem.type,
-      };
+
+  setVariable: async function ( scope, { field, var_data, from_table }) {
+    /* Returns true if the field was set.
+    *
+    * TODO: Will the selector ever change the second time around?
+    *   If so, we need to store the handles before this.
+    * 
+    * @returns bool
+   */
+    let was_set = false;
+
+    let matching_input_rows = await scope.getMatchingRows( scope, { var_data, field, from_table });
+
+    // Nothing to set if there are no matching table rows
+    if ( matching_input_rows.length <= 0 ) { return was_set; }
+    // Want this info for the report later
+    field.matching_input_rows = matching_input_rows;
+
+    // Take a while to load sometimes
+    if ( field.tag === 'canvas' ) {
+      await scope.page.waitForSelector('#dasigcanvas');
+    }
+
+    let handle = await scope.page.evaluateHandle(( selector, field ) => {
+      let elem = document.querySelector( selector );
+      // `elem` has never been null
+      if ( elem.nextSibling && elem.nextSibling.tag === `label` ) {
+        return elem.nextSibling;
+      } else { return elem; }
+    }, field.selector, field );
+
+    let disabled = await handle.evaluate(( elem )=> {
+      // da's header covers the top of the section containing the fields which means
+      // sometimes it blocks nodes from clicks. Scroll item to the bottom instead.
+      if ( !elem.disabled ) { elem.scrollIntoView(false); }
+      return elem.disabled;
     });
 
     // If field is disabled, don't set it yet
-    if ( disabled ) { return !disabled; }
+    if ( disabled ) { return was_set; }
 
-    // da's header covers the top of the section containing the fields which means
-    // sometimes it blocks nodes from clicks. Scroll item to the bottom instead.
-    await handle.evaluate(( elem ) => { elem.scrollIntoView(false); return elem.scrollTop });
-    let set_to = row.value;
-
-    if ( env_vars.DEBUG && handle ) {
-      let name = await handle.evaluate(( elem ) => {
+    if ( env_vars.DEBUG ) {
+      let name = await handle.evaluate(( elem ) => {  // Not using this right now
         try { elem.nextSibling.style.background = 'Khaki'; }
         catch(err){ elem.style.background = 'Khaki'; }
         return elem.getAttribute('for') || elem.getAttribute('name'); 
       });
       await scope.page.waitFor( 400 );
     }
 
+    let { tag, type } = field;
+    // We use the last match. We wrote a warning for the user if there were
+    // mutliple matches in the story that the last value will be used. In future,
+    // Steps will be able to set this data gradually and we don't want the user
+    // to get confused thinking that an earlier Step was used instead of a later one.
+    // TODO: Add the `was_used` prop to the row so it can be used in the report.
+    let row = matching_input_rows[ matching_input_rows.length - 1 ];
+    let set_to = row.value;
```

Comment:
For `there_is_another`, this is usually a button, so it shouldn't usually matter, but this is not actually always `True` or `False` - the value needs to get mutated or something. And then mutated back??? (from the original row) We want to keep the list of rows referring to the original rows, right? Maybe we don't have to, though, if we collect used rows separately.

Reply:
Send back the right row values from `matches`, but from `setVariable` send back the original row that was used...? Is that possible? Maybe the fake row can have a reference to the original row in it?